airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialLoadHandler.java

```diff
@@ -0,0 +1,224 @@
+  }
+
+  public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(
+      final ConfiguredAirbyteCatalog catalog,
+      final Map<String, TableInfo<CommonField<MysqlType>>> tableNameToTable,
+      final Instant emittedAt) {
+    final List<AutoCloseableIterator<AirbyteMessage>> iteratorList = new ArrayList<>();
+    for (final ConfiguredAirbyteStream airbyteStream : catalog.getStreams()) {
+      final AirbyteStream stream = airbyteStream.getStream();
+      final String streamName = stream.getName();
+      final String namespace = stream.getNamespace();
+      final AirbyteStreamNameNamespacePair pair = new AirbyteStreamNameNamespacePair(streamName, namespace);
+      final String fullyQualifiedTableName = DbSourceDiscoverUtil.getFullyQualifiedTableName(namespace, streamName);
+      if (!tableNameToTable.containsKey(fullyQualifiedTableName)) {
+        LOGGER.info("Skipping stream {} because it is not in the source", fullyQualifiedTableName);
+        continue;
+      }
+      if (airbyteStream.getSyncMode().equals(SyncMode.INCREMENTAL)) {
+        // Grab the selected fields to sync
+        final TableInfo<CommonField<MysqlType>> table = tableNameToTable
+            .get(fullyQualifiedTableName);
+        final List<String> selectedDatabaseFields = table.getFields()
+            .stream()
+            .map(CommonField::getName)
+            .filter(CatalogHelpers.getTopLevelFieldNames(airbyteStream)::contains)
+            .collect(Collectors.toList());
+        final AutoCloseableIterator<JsonNode> queryStream = queryTablePk(selectedDatabaseFields, table.getNameSpace(), table.getName());
+        final AutoCloseableIterator<AirbyteMessage> recordIterator =
+            getRecordIterator(queryStream, streamName, namespace, emittedAt.toEpochMilli());
+        final AutoCloseableIterator<AirbyteMessage> recordAndMessageIterator = augmentWithState(recordIterator, pair);
+
+        iteratorList.add(augmentWithLogs(recordAndMessageIterator, pair, streamName));
+
+      }
+    }
+    return iteratorList;
+  }
+
+  private AutoCloseableIterator<JsonNode> queryTablePk(
+      final List<String> columnNames,
+      final String schemaName,
+      final String tableName) {
+    LOGGER.info("Queueing query for table: {}", tableName);
+    final AirbyteStreamNameNamespacePair airbyteStream =
+        AirbyteStreamUtils.convertFromNameAndNamespace(tableName, schemaName);
+    return AutoCloseableIterators.lazyIterator(() -> {
+      try {
+        final Stream<JsonNode> stream = database.unsafeQuery(
+            connection -> createPkQueryStatement(connection, columnNames, schemaName, tableName, airbyteStream),
+            sourceOperations::rowToJson);
+        return AutoCloseableIterators.fromStream(stream, airbyteStream);
+      } catch (final SQLException e) {
+        throw new RuntimeException(e);
+      }
+    }, airbyteStream);
+  }
+
+  private PreparedStatement createPkQueryStatement(
```

Comment:
not a code comment but an observation that i thought we could reuse code from `AbstractJdbcSource.java`  and pass a `CursorInfo` object that contained the primary key column name and an optional checkpoint value. Is this new code path very different? https://github.com/airbytehq/airbyte/blob/ac67bfd9fdde0a4a02ed86f2508423fc50c187d5/airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java#L315 source to query.

Reply:
That's a good point - this is super similar, but there are some subtle differences : 
1. Since the primary key is unique, we don't need the logic to determine whether we need to do a >= or > query
2. We will always be ordering in ASC order

I think think in the future we can certainly refactor this, but for now I think it's easier to keep it seperate